teps/0018-default-bundle.md

```diff
@@ -0,0 +1,165 @@
+
+## Requirements
+
+## Proposal
+
+Create a new key for PipelineRun and TaskRun called `spec.lookup.bundle`; use it
+for finding referenced names in Tasks and Pipelines.
+
+### User Stories
+
+#### Story 1
+
+I have a repository where I have the entire CI/CD system using Tekton
+resources defined in the repository. I wish to test commits to the repository
+(presubmits) before allowing them to merge. I wish to support using the exact
+resources when running locally.
+
+My Pipelines refer to Tasks defined in-repo by name only and same for
+PipelineRun->Pipeline and TaskRun->Task. I can still use `_Ref.bundle` if I
+am referring to resources which are not created by this branch of the repository.
+
+My presubmit test takes non-Run resources and creates a bundle. All PipelineRuns
+and TaskRuns are modified to use the bundle as the default lookup environment. I
+can now execute in the production namespace without interfering with the
+environment in the existing namespace.
+
+#### Story 2
+
+I wish to support multiple repositories with CI/CD defined in-repo in the same
+namespace without the possibility of name conflicts. The system creating
+the PipelineRun must always specify a default bundle.
+
+## Detailed Design
+
+PipelineRun and TaskRun will be modified to support `spec.lookup.bundle`.
+
+```go
+// Lookup specifies where to find references by name
+// +optional
+Lookup Lookup `json:"lookup,omitempty"`
+...
+type Lookup struct {
+    // Bundle url reference to a Tekton Bundle
+    // +optional
+    Bundle string `json:"bundle,omitempty"`
+}
+```
+
+If `spec.lookup.bundle` is specified, the local database is never queried to
+find a bare `taskRef.name` or `pipelineRef.name`.
+
+### Observability
+
+TBD need advice
+
+### New Variables
+
+In Pipeline: `context.pipelineRun.lookupBundle`
+
+In Task: `context.taskRun.lookupBundle`
+
+Definition: Where the _Run is finding references by name if not the local database.
+
+### Explicit Bundles from in-repo Resources
+
+If any Tekton resources specify `bundle:` referring to bundles generated from
+resources specified in the repository being modified, a presubmit test must be
+aware of it and must modify the bundle reference to point to a version generated
+from the current commit being tested.
+
+## Test Plan
+
+Unit tests for bundles which cannot be accessed, bundles which contain some but
+not all referenced resources, and bundles which have all referenced resources.
+
+An integration test with a Pipeline referencing a Task by name which is not
+located in the cluster but only in the bundle.
+
+## Drawbacks
+
+New variables have very narrow uses.
+
+Could be confusing when two PipelineRuns for the same Pipeline running
+simultaneously can be executing different Tasks.
+
+## Alternatives
+
+1. All references could be required to have a bundle key. When you want to use
+a custom version locally, you will need to modify the bundle to point to your
+local copy. Tooling could make this trivial to do.
```

Comment:
Ah, I should have read the alternatives, my comments are describing this alternative I guess.

Reply:
:-)